### PR TITLE
Refactor bundle.go logic

### PR DESCRIPTION
**Pull request by usernamedt**
I propose decomposing complex bundle.go logic into several structs as shown on the diagram below:
 image

Now bundle.go uses TarBallComposer to create tars from files encountered during filesystem walk. We can change TarBallComposer implementation, as for this PR there is only RegularTarBallComposer.
Also, each tar ball now has its own name. TarFileSets now holds information about the contents of each tarball.

### CODE REVIEW

File: internal/tar_ball_queue.go [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+		panic("Trying to deque from not started Queue")
+	}
+	return <-tarQueue.tarballQueue
+}
+
+func (tarQueue *TarBallQueue) FinishQueue() error {
+	if !tarQueue.started {
+		panic("Trying to stop not started Queue")
+	}
+	tarQueue.started = false
```
Comment (g0djan) [inline internal/tar_ball_queue.go:68]: which methods could be invoked in a goroutine? I see you access this variable in several places and it looks like there is no synchronization to access it.

File: internal/tar_ball_queue.go [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+		AllTarballsSize:  new(int64),
+	}
+}
+
+func (tarQueue *TarBallQueue) StartQueue() error {
+	if tarQueue.started {
+		panic("Trying to start already started Queue")
+	}
+	var err error
+	tarQueue.parallelTarballs, err = getMaxUploadDiskConcurrency()
```
[the remark this answers is not in the document]
Comment (usernamedt) [reply internal/tar_ball_queue.go:37]: I didn't change the original code here, just decomposed it into different structs.

File: internal/tar_ball_file_packer.go [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+			return nil
+		} else if err != nil {
+			return errors.Wrapf(err, "PackFileIntoTar: failed to find corresponding bitmap '%s'\n", cfi.path)
+		}
+		fileReader, cfi.header.Size, err = ReadIncrementalFile(cfi.path, cfi.fileInfo.Size(), *incrementBaseLsn, bitmap)
+		if os.IsNotExist(err) { // File was deleted before opening
+			// We should ignore file here as if it did not exist.
+			return nil
+		}
+		switch err.(type) {
```
[the remark this answers is not in the document]
Comment (usernamedt) [reply internal/tar_ball_file_packer.go:57]: I didn't change the original code here, just decomposed it into different structs.

File: internal/tar_ball_queue.go (hunk 1 above)
Comment (usernamedt) [reply internal/tar_ball_queue.go:68]: I can add this information in methods documentation

File: internal/bundle_file_list.go [hunk 4]
```diff
@@ -0,0 +1,32 @@
+package internal
+
+import (
+	"archive/tar"
+	"os"
+	"sync"
+)
+
+// BundleFileList represents the files in the backup that is going to be created
+type BundleFileList interface {
```
[the remark this answers is not in the document]
Comment (usernamedt) [reply internal/bundle_file_list.go]: Renamed in 2ee0d3a

File: internal/tar_ball_composer.go [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+	RegularComposer TarBallComposerType = iota + 1
+)
+
+func NewTarBallComposer(composerType TarBallComposerType, bundle *Bundle) (TarBallComposer, error) {
+	switch composerType {
+	case RegularComposer:
+		fileList := &RegularBundleFileList{}
+		tarBallFilePacker := newTarBallFilePacker(bundle.DeltaMap, bundle.IncrementFromLsn, fileList)
+		return NewRegularTarBallComposer(bundle.TarBallQueue, tarBallFilePacker, fileList, bundle.Crypter), nil
+	default:
```
[the remark this answers is not in the document]
Comment (usernamedt) [reply internal/tar_ball_composer.go:47]: Fixed in 08a2ad7

File: internal/tar_ball_queue.go [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+
+import (
+	"github.com/pkg/errors"
+	"sync"
+	"sync/atomic"
+)
+
+// TarBallQueue is used to process multiple tarballs concurrently
+type TarBallQueue struct {
+	tarballQueue     chan TarBall
```
[the remark this answers is not in the document]
Comment (usernamedt) [reply internal/tar_ball_queue.go]: Fixed in 1088312

File: internal/tar_ball_queue.go (hunk 1 above)
Comment (usernamedt) [reply internal/tar_ball_queue.go]: I've replaced default bool with atomic bool in 2d464f4

**Pull request merged**